rust/parser/mod.rs

```diff
@@ -322,14 +337,14 @@ fn visit_pattern_variable(ctx: Rc<Pattern_variableContext>) -> ParserResult<Vari
     }
 }
 
-fn visit_variable_concept(_ctx: Rc<Variable_conceptContext>) -> ParserResult<TypeVariable> {
-    todo!()
+fn visit_variable_concept(ctx: Rc<Variable_conceptContext>) -> ParserResult<ConceptVariable> {
+    get_var(ctx.VAR_(0).unwrap()).is(get_var(ctx.VAR_(1).unwrap()))
 }
 
 fn visit_variable_type(ctx: Rc<Variable_typeContext>) -> ParserResult<TypeVariable> {
     let mut var_type = match visit_type_any(ctx.type_any().unwrap())? {
         Type::Variable(p) => p,
-        Type::Label(p) => UnboundVariable::hidden().type_(p)?.into_type(),
+        Type::Label(p) => UnboundVariable::hidden().type_(p)?,
```

Comment:
The tightened type system doesn't require wrapping into and unwrapping from the enums. Most `into_X()` invocations are made redundant as a result.

Reply:
Do you mean things like `UnboundVariable::into_thing()` or `SomeConstraint::from::<UnboundVariable>()`? Both are needed, because the meaning behind a variable being unbound is that we don't yet know what kind of variable it is, and its context will help determine whether it's a thing, a type, or a concept.

Methods like `into_variable()` and `into_pattern()` are still needed for actual type erasure so that we can store all the different kinds of patterns in a homogenous collection, e.g. `Conjunction` contains a `Vec<Pattern>`.